ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,127 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "repository" end)
+
+open D
+
+let cmd_yum = "/usr/bin/yum"
+let cmd_rm = "/usr/bin/rm"
+
+let reposync_mutex = Mutex.create ()
+
+let create_repository_record ~__context ~name_label ~name_description ~binary_url ~source_url =
+  let ref = Ref.make () in
+  let uuid = Uuidm.to_string (Uuidm.create `V4) in
+  Db.Repository.create
+    ~__context
+    ~ref
+    ~uuid
+    ~name_label
+    ~name_description
+    ~binary_url
+    ~source_url
+    ~hash:""
+    ~up_to_date:false;
+  ref
+
+let assert_name_is_valid ~name_label =
+  match Re.Str.string_match (Re.Str.regexp "^[A-Za-z]+[A-Za-z0-9\\-]+$") name_label 0 with
+  | true -> ()
+  | false -> raise Api_errors.(Server_error (invalid_repository_name, [name_label]))
+
+let assert_url_is_valid ~url =
+  match !Xapi_globs.repository_domain_name_allowlist with
+  | [] -> ()
+  | l ->
+    let result = l |> List.exists (fun domain_name ->
```

Comment:
Can we use `Uri.of_string` (see e.g. `guest_agent_config_requirements` in xapi_pool.ml)?

Reply:
I can't see any benefit from `Uri.of_string` and `Uri.scheme` as I can' find how these functions check the URL.
Please correct me if I missed something by reading https://github.com/mirage/ocaml-uri/blob/master/lib/uri.ml.

Reply:
I assume that `Uri.of_string` returns an error if it fails to parse the given string as a well-formed URL. We can use it to sanity check the whole URL. You still need to check whether the domain is in the allowlist. I think we can use `Uri.host` and match the suffix.

I find it hard to verify that the regexp below that all it needs to do (but perhaps it's just me :)).